python/ray/serve/_private/deployment_state.py

```diff
@@ -2374,8 +2494,210 @@ def check_and_update_replicas(self):
                 self._replicas.add(ReplicaState.STOPPING, replica)
             else:
                 logger.info(f"{replica.replica_id} is stopped.")
+
+                # Release rank only after replica is successfully stopped
+                # This ensures rank is available during draining/graceful shutdown
+                replica_id = replica.replica_id.unique_id
+                self.release_replica_rank(replica_id)
+
                 self._autoscaling_state_manager.on_replica_stopped(replica.replica_id)
 
+        # After replica state updates, check rank consistency and perform minimal reassignment if needed
+        # This ensures ranks are continuous after lifecycle events
+        # Only do consistency check when deployment is stable (not during active updates)
+        # maybe this constraint need to be relaxed in the future. The implication is that
+        # if we delay the rank reassignment, the rank system will be in an invalid state
+        # for a longer period of time. Abrar made this decision because he is not confident
+        # about how rollouts work in the deployment state machine.
+        if not self._has_outdated_version_replicas():
+            self._check_rank_consistency_and_reassign_minimally()
```

Comment:
It seems like the code could be more clean if we split this logic out into a `ReplicaRankManager`? we can call into it with `rank_manager.check_rank_consistency_and_reassign_minimally()`, `rank_manager.release_replica_rank()`, `rank_manager.get_next_available_rank()` etc. Then at the end of the deployment state update, we can call reconfigure on all actors according to the rank mapping kept in rank manager -- wdyt?

Reply:
One thing to keep in mind: there's now two callsites of `replica.reconfigure`. I think this is acceptable for now, but let's make sure this doesn't cause any issues then add a note/todo about it in the codebase.